mypy/applytype.py

```diff
@@ -43,6 +43,10 @@ def apply_generic_arguments(callable: CallableType, types: List[Type],
             else:
                 msg.incompatible_typevar_value(callable, i + 1, type, context)
 
+        # Translate constructor fn back to class.
+        if isinstance(type, CallableType) and type.is_type_obj():
+            types[i] = type = type.ret_type
+
```

Comment:
I don't think this is right, though somehow it never made a difference in my tests. The value `C` (the class) is not the same sort of thing as an instance of the class `C`.